hs_core/hydroshare/resource.py

```diff
@@ -29,16 +30,36 @@
 logger = logging.getLogger(__name__)
 
 
-def update_quota_usage(res):
-    from hs_core.tasks import update_quota_usage_task
-    quser = res.get_quota_holder()
-    if quser is None:
-        # no quota holder for this resource, this should not happen, but check just in case
-        logger.error('no quota holder is found for resource' + res.short_id)
+def update_quota_usage(res=None, user=None):
+    """
+    Update quota usage as follows:
+    (1) if res is not None & user is None, quota usage of the res' quota holder will be updated
+    (2) if res is None & user is not None, quota usage of the user will be updated
+    (3) if both res and user are None, nothing is updated
+    (4) if both res and user are not None, quota usage of both res' quota holder and user will be
+    updated.
+    :param res: a resource object
+    :param user: a user object
+    :return:
+    """
+    if not res and not user:
```

Comment:
what happens here if res is set but not user. Could it be `if not res or not user`?

Reply:
@aphelionz If res is set but not user, the quota of the quota holder of that res will be updated by this routine. If res is not set but user is set, the quota of the set user will be updated, which is needed for some scenario such as when replicating a public resource bag to a login user's user zone. If both res and user are set, quota of both res' quota holder and the set user will be updated. If both are not set, this routine just returns without doing anything since no input to update quota. The details are included in the docstring of this routine. The two input parameters are needed to tailor to different scenarios, including how iRODS quota micro-service is implemented which depends on quota holder to be set on a resource collection before adding files to the collection, otherwise, added files will not be counted toward the quota by iRODS quota micro-service. I have debugged and tested various scenarios and made this routine the single point of truth for updating quota usage. Let me know if this does not clarify the confusion.

Reply:
Thank you!